core/app/models/spree/payment_source.rb

```diff
@@ -23,9 +23,9 @@ def can_capture?(payment)
     end
 
     # @param payment [Spree::Payment] the payment we want to know if can be voided
-    # @return [Boolean] true when the payment is not failed or voided
+    # @return [Boolean] true when the payment is not failed, voided or invalid
     def can_void?(payment)
-      !payment.failed? && !payment.void?
+      !payment.failed? && !payment.void? && !payment.invalid?
```

Comment:
Have you considered using `payment.can_void?`. It should just use a [metaprogrammed `state_machine` helper](https://github.com/state-machines/state_machines/blob/5cdb6783edd4e14323276bdb14fb5bcac077d469/lib/state_machines/event.rb#L209) that should return the same value since the only allowed states would be the remaining ones as you linked in the PR description.

Reply:
Yeah, it should have the same behaviour. I'll update the PR and remove both tests, given that the logic will not be in payment source anymore and this behaviour is tested in other place.
Thanks for the review!